Combine terraform docs for GCP and Azure

sgibson91 opened the pull request:
This PR adds documentation on how to deploy a cluster to Azure using terraform. Since this overlaps very closely with the GCP workflow, I've opted to use panels for the provider-specific sections so we can keep all the docs in one place and maintain them more easily.

View a rendered version here: https://2i2c-pilot-hubs--836.org.readthedocs.build/en/836/howto/operate/new-cluster/new-cluster.html

fixes https://github.com/2i2c-org/infrastructure/issues/718

yuvipanda commented on docs/howto/operate/new-cluster/azure-or-google-cloud.md:
`````diff
@@ -26,8 +31,47 @@ Example `.tfvars` file:
 ```
+````
+
+````{tabbed} Azure
`````
Nice!

yuvipanda commented on docs/howto/operate/new-cluster/azure-or-google-cloud.md:
```diff
@@ -26,8 +31,47 @@ Example `.tfvars` file:
+- `subscription_id`: Azure subscription ID to create resources in.
+  Should be the id, rather than display name of the project.
+- `resourcegroup_name`: The name of the Resource Group that the cluster and other resources will be deployed into.
+- `global_container_registry_name`: The name of an Azure Container Registry to use for our image.
```
Let's provide guidelines on naming conventions here? I'm not sure what the current names are, we can just pick that and riff off that?

yuvipanda commented on docs/howto/operate/new-cluster/azure-or-google-cloud.md:
````diff
@@ -26,8 +31,47 @@ Example `.tfvars` file:
+  az acr check-name --name ACR_NAME --output table
+  ```
+
+- `global_storage_account_name`: The name of a storage account to use for Azure File Storage.
````
Let's provide guidelines on naming conventions here? I'm not sure what the current names are, we can just pick that and riff off that?

yuvipanda reviewed (approved):
One comment, but otherwise this is pretty cool!

sgibson91 commented:
@yuvipanda RE: Naming conventions (replying here since the two threads are very similar)

This is what JIL has:

https://github.com/2i2c-org/infrastructure/blob/da84ec39e603a801c0a5f7cf8605bba59500dec1/terraform/azure/projects/justiceinnovationlab.tfvars#L7-L8

And this is what I did for Carbon Plan:

https://github.com/2i2c-org/infrastructure/blob/da84ec39e603a801c0a5f7cf8605bba59500dec1/terraform/azure/projects/carbonplan.tfvars#L6-L7

I think I'd like to put my vote behind the `{CLUSTER_NAME}registry`/`{CLUSTER_NAME}storage` approach as it's more readable. And maybe we also add in a `lowerCamelCase` guideline too? so `carbonplanstorage` becomes `carbonplanStorage`?

yuvipanda commented:
> I think I'd like to put my vote behind the `{CLUSTER_NAME}registry`/`{CLUSTER_NAME}storage` approach as it's more readable. And maybe we also add in a `lowerCamelCase` guideline too? so `carbonplanstorage` becomes `carbonplanStorage`?

happy with that, although I want at least some prefix / suffix indicating this is for a JupyterHub (not necessarily 2i2c) - carbonplan probably has *other* infra they deploy, and I don't want to take that namespace up. +1 on camelCase if azure supports it - I remember it had super restrictive naming rules.

sgibson91 commented:
Ok, so `{CLUSTER_NAME}HubRegistry`? and equivalent for storage?

> +1 on camelCase if azure supports it - I remember it had super restrictive naming rules.

Camel case is fine. Restrictions are: alphanumeric between 5 and 50 characters in length for container registries, and alphanumeric between 3-24 characters for storage accounts.

yuvipanda commented:
Sounds good, @sgibson91

sgibson91 commented:
Ok @yuvipanda, I pushed a commit, LMK what you think. Link to rendered version if it's easier: https://2i2c-pilot-hubs--836.org.readthedocs.build/en/836/howto/operate/new-cluster/azure-or-google-cloud.html

sgibson91 commented on docs/howto/operate/new-cluster/azure-or-google-cloud.md:
```diff
@@ -0,0 +1,198 @@
+
+Names for Azure container registries and storage accounts **must** conform to the following guidelines:
+
+- alphanumeric strings between 5 and 50 characters long for [container registries](https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/resource-name-rules#microsoftcontainerregistry), e.g., `myContainerRegistry007`
```
```suggestion
- alphanumeric strings between 5 and 50 characters for [container registries](https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/resource-name-rules#microsoftcontainerregistry), e.g., `myContainerRegistry007`
```

choldgraf commented on docs/requirements.txt:
```diff
@@ -4,4 +4,5 @@ pyyaml
 requests
 sphinx-autobuild
 sphinx-panels
+sphinx-tabs
```
I think we could re-use the functionality that is in sphinx-panels for this, rather than adding a new dependency: https://sphinx-panels.readthedocs.io/en/latest/#tabbed-content

sgibson91 replied:
Ah sorry, I'm still super new to using Sphinx directly (as opposed to Jupyter Book). If we have capacity, I'd love a sphinx directive cheatsheet for common things we include in our docs, maybe keep it in the team-compass? Another thing that **always** catches me out is that examples in the sphinx docs are usually in rST and I end up staring at my broken Markdown build for hours wondering what the problem is because I copy-pasta'd.

sgibson91 replied:
Done in [`d449468`](https://github.com/2i2c-org/infrastructure/pull/836/commits/d449468536dfdd9afeca00a49777ec4f0ab71e57)

choldgraf replied:
@sgibson91 that's a cool idea re: common directives. Maybe we can add something to the team compass that just has a list of stuff that is useful?

sgibson91 replied:
I would love that!

choldgraf commented on docs/howto/operate/new-cluster/azure-or-google-cloud.md:
```diff
@@ -0,0 +1,198 @@
+# Add a new Azure or Google Cloud cluster
```
If we imagine the AWS deployment information to be very similar, maybe we should just leave this as `Add a new Kubernetes Cluster`  and not have it name specific clouds. And we can add a `TODO` item to fill out the AWS-specific information.

sgibson91 replied:
Done across [`7e7fbe0`](https://github.com/2i2c-org/infrastructure/pull/836/commits/7e7fbe051ad619158a6e235d4957faa97ac86d80) and [`26c3b2b`](https://github.com/2i2c-org/infrastructure/pull/836/commits/26c3b2b484ff9afdb3ff1dc4de193c6184361682)

choldgraf reviewed (commented):
I'll take a more in-depth pass later, but just wanted to make two quick comments. This looks great, I love the idea of standardizing our deployment details across providers.

sgibson91 commented on docs/howto/operate/new-cluster/new-cluster.md:
```diff
@@ -0,0 +1,204 @@
+
+Congratulations, you've just deployed a new cluster!
+
+## Exporting and Encrypting the Continuous Deployment Service Account
```
```suggestion
## Exporting and Encrypting the Cluster Access Credentials
```

choldgraf commented on docs/howto/operate/new-cluster/new-cluster.md:
```diff
@@ -0,0 +1,204 @@
+
+This guide will assume you have already followed the guidance in [](/topic/cluster-design) to select the appropriate infrastructure.
+
+## Creating a Terraform variables file for the cluster
```
```suggestion
## Create a Terraform variables file for the cluster
```

I think in general we go with "imperative" (create) over gerund-y verbs (creating)...not super opinionated about that though

choldgraf commented on docs/howto/operate/new-cluster/new-cluster.md:
````diff
@@ -0,0 +1,204 @@
+  You can use the following command to check your desired name is available:
+
+  ```bash
+  az acr check-name --name ACR_NAME --output table
````
Can we link to some docs about what `az` is here?

choldgraf commented on docs/howto/operate/new-cluster/new-cluster.md:
```diff
@@ -0,0 +1,204 @@
+(new-cluster:new-cluster)=
+# Add a new Kubernetes cluster
+
+This guide will walk through the process of adding a new cluster to our [terraform configuration](https://github.com/2i2c-org/infrastructure/tree/HEAD/terraform).
```
Could we cross-link to a section in the docs that explains what terraform is and links to docs? (I think this page already exists somewhere, if it doesn't then creating that page doesn't need to be part of this PR)

choldgraf commented on docs/howto/operate/new-cluster/new-cluster.md:
```diff
@@ -0,0 +1,204 @@
+
+## Initialising Terraform
+
+Our default terraform state is located centrally in our `two-eye-two-see-org` GCP project, therefore you must authenticate `gcloud` to your `@2i2c.org` account before initialising terraform.
```
```suggestion
Our default terraform state is located centrally in our `two-eye-two-see-org` GCP project, therefore you must authenticate `gcloud` to your `@2i2c.org` account before initialising terraform.
The terraform state includes **all** cloud providers, not just GCP.
```

choldgraf commented on docs/howto/operate/new-cluster/new-cluster.md:
````diff
@@ -0,0 +1,204 @@
+## Plan and Apply Changes
+
+```{note}
+When deploying to Google Cloud, make sure the [Artifact Registry API](https://console.cloud.google.com/apis/library/artifactregistry.googleapis.com) in enabled on the project before deploying!
````
```suggestion
When deploying to Google Cloud, make sure the [Artifact Registry API](https://console.cloud.google.com/apis/library/artifactregistry.googleapis.com) is enabled on the project before deploying!
```

choldgraf commented on docs/howto/operate/new-cluster/new-cluster.md:
````diff
@@ -0,0 +1,204 @@
+terraform plan -var-file=projects/CLUSTER.tfvars
+```
+
+Check over the output of this command to ensure nothing if being created/deleted than you expected.
````
```suggestion
Check over the output of this command to ensure nothing is being created/deleted that you didn't expect.
```

choldgraf reviewed (commented):
This looks nicely written to me! I left a few comments in there, but none were very major.

I wonder if it would be useful for somebody on the team to follow these instructions from start to finish, to see if there is any missing piece in there. Then again, I suspect we'll *need* to follow the instructions in the future anyway, and can bring up ideas for improvement then. What do you think?

sgibson91 commented:
> I wonder if it would be useful for somebody on the team to follow these instructions from start to finish, to see if there is any missing piece in there. Then again, I suspect we'll _need_ to follow the instructions in the future anyway, and can bring up ideas for improvement then. What do you think?

I'm +1 on merging and letting other folk raise issues when we next deploy a cluster. While I have mostly worked on these docs myself, I've revisited them in 3 very different situations: simple deployment for COESSING, complex deployment for Pangeo, totally new deployment for Azure. So I think we have a lot of gotchas in there already.

sgibson91 commented:
I've addressed all the comments I got and cleaned up the language on our terraform page when I added the cross-ref, as that was still very GCP-specific. So I'm going to merge this now :)
